### PR TITLE
refactor(components): [input] refactor

### PR DESCRIPTION
**Pull request by sxzz**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Review by HerringtonDarkholme — APPROVED** [review]

**Comment by HerringtonDarkholme** [comment]
Please take a look at the test. The test looks broken by input change.

File: packages/components/autocomplete/src/autocomplete.vue [hunk 1]
```diff
@@ -280,10 +278,11 @@ const highlight = (index: number) => {
...
 onMounted(() => {
-  inputRef.value!.inputOrTextarea.setAttribute('role', 'textbox')
-  inputRef.value!.inputOrTextarea.setAttribute('aria-autocomplete', 'list')
-  inputRef.value!.inputOrTextarea.setAttribute('aria-controls', 'id')
-  inputRef.value!.inputOrTextarea.setAttribute(
+  // TODO: use Volar generate dts to fix it.
+  ;(inputRef.value as any).ref!.setAttribute('role', 'textbox')
+  ;(inputRef.value as any).ref!.setAttribute('aria-autocomplete', 'list')
+  ;(inputRef.value as any).ref!.setAttribute('aria-controls', 'id')
+  ;(inputRef.value as any).ref!.setAttribute(
```
Comment (JeremyWuuuuu) [inline packages/components/autocomplete/src/autocomplete.vue:285]: **Suggestion:**

```ts
const inputElement = (inputRef.value as any).ref!
[
  { key: 'role', value: 'textbox'},
  { key: 'aria-autocomplete', value: 'list' },
  { key: 'aria-controls', value: 'id' },
  { key: 'aria-activedescendant', value: `${id.value}-item-${highlightedIndex.value}` }
].forEach(({ key, value }) => inputElement.setAttribute(key, value))

```

File: packages/components/input/src/input.vue [hunk 2]
```diff
@@ -162,336 +165,304 @@ import {
...
+  // hack for https://github.com/ElemeFE/element/issues/8548
+  // should remove the following line when we don't support IE
+  if (value === nativeInputValue.value) return
+
+  emit(UPDATE_MODEL_EVENT, value)
+  emit('input', value)
+
+  // ensure native input value is controlled
+  // see: https://github.com/ElemeFE/element/issues/12850
+  nextTick(setNativeInputValue)
```
Comment (JeremyWuuuuu) [inline packages/components/input/src/input.vue:328]: **Nit** 
```suggestion
  nextTick(() => setNativeInputValue())
```

File: packages/components/input/src/input.vue [hunk 3]
```diff
@@ -162,336 +165,304 @@ import {
...
+  emit(UPDATE_MODEL_EVENT, '')
+  emit('change', '')
+  emit('clear')
+  emit('input', '')
+}
+
+watch(
+  () => props.modelValue,
+  () => {
+    nextTick(resizeTextarea)
```
Comment (JeremyWuuuuu) [inline packages/components/input/src/input.vue:409]: **Nit**

```suggestion
    nextTick(() => resizeTextarea())
```

File: packages/components/input/src/input.vue [hunk 4]
```diff
@@ -162,336 +165,304 @@ import {
...
+    setNativeInputValue()
+    resizeTextarea()
+    updateIconOffset()
+  }
+)
+
+onMounted(() => {
+  setNativeInputValue()
+  updateIconOffset()
+  nextTick(resizeTextarea)
```
Comment (JeremyWuuuuu) [inline packages/components/input/src/input.vue:437]: **Nit**
```suggestion
  nextTick(() => resizeTextarea)
```

File: packages/components/input/src/input.vue [hunk 5]
```diff
@@ -162,336 +165,304 @@ import {
...
+})
 
-    const handleKeydown = (evt: KeyboardEvent) => {
-      emit('keydown', evt)
-    }
+onUpdated(() => {
+  nextTick(updateIconOffset)
```
Comment (JeremyWuuuuu) [inline packages/components/input/src/input.vue:441]:
```suggestion
  nextTick(() => updateIconOffset())
```

**Review by jw-foss — COMMENTED** [review]
Left a couple comments for you to consider 🏓

File: packages/components/autocomplete/src/autocomplete.vue (hunk 1 above)
Comment (sxzz) [reply packages/components/autocomplete/src/autocomplete.vue:285]: image

This is not the main refactoring component of this PR, and we can do it in another PR.

**Comment by sxzz** [comment]
@JeremyWuuuuu 🏓 Done!

**Review by jw-foss — APPROVED** [review]

**Pull request merged**